From 7104731a26a4124cc22f512b816c6f3a648bcf25 Mon Sep 17 00:00:00 2001
From: Lukas Czerner <lczerner@redhat.com>
Date: Thu, 23 Oct 2014 09:54:05 -0400
Subject: [fs] ext4: fix bigalloc regression

Message-id: <1414058125-4183-92-git-send-email-lczerner@redhat.com>
Patchwork-id: 98192
O-Subject: [RHEL 7.1 PATCH v2 091/171] ext4: fix bigalloc regression
Bugzilla: 1150139
RH-Acked-by: Eric Sandeen <sandeen@redhat.com>

From: Eric Whitney <enwlinux@gmail.com>

BZ 1150139
https://bugzilla.redhat.com/show_bug.cgi?id=1150139

BREW 8142710
https://brewweb.devel.redhat.com/taskinfo?taskID=8142710

Upstream commit d0abafac8c9162f39c4f6b2f8141b772a09b3770

 Commit f5a44db5d2 introduced a regression on filesystems created with
 the bigalloc feature (cluster size > blocksize).  It causes xfstests
 generic/006 and /013 to fail with an unexpected JBD2 failure and
 transaction abort that leaves the test file system in a read only state.
 Other xfstests run on bigalloc file systems are likely to fail as well.

 The cause is the accidental use of a cluster mask where a cluster
 offset was needed in ext4_ext_map_blocks().

 Signed-off-by: Eric Whitney <enwlinux@gmail.com>
 Signed-off-by: Lukas Czerner <lczerner@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/ext4/extents.c b/fs/ext4/extents.c
index ce94edc..87bd1b9 100644
--- a/fs/ext4/extents.c
+++ b/fs/ext4/extents.c
@@ -4218,7 +4218,7 @@ int ext4_ext_map_blocks(handle_t *handle, struct inode *inode,
   */
  map->m_flags &= ~EXT4_MAP_FROM_CLUSTER;
  newex.ee_block = cpu_to_le32(map->m_lblk);
- cluster_offset = EXT4_LBLK_CMASK(sbi, map->m_lblk);
+ cluster_offset = EXT4_LBLK_COFF(sbi, map->m_lblk);
 
  /*
   * If we are doing bigalloc, check to see if the extent returned
-- 
1.7.1